test: fixed-length strings in netvars

mattwalsh-unity opened the pull request:
Just a sketch with which I seriously can use some help.

I wanted to prove to myself that I could do fixed strings w/ network variables, and this really weak example is where I ended up.  I need to learn more about how to do this more properly with the `fixed` statements in c# and/or `FixedString`

mattwalsh-unity committed:
horrible fixed string experiment

mattwalsh-unity committed:
horrible fixed string experiment

ShadauxCat commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkVariableTests.cs:
```diff
@@ -4,8 +4,63 @@
 {
+    public unsafe struct FixedStringStruct : INetworkSerializable
+    {
+        public fixed char FixedString[6];
```
I would make this a `public unsafe char*`  instead of `public fixed char[]`. I think even with `fixed` this will cause a GC alloc... just a pinned one. With `char*` you can use `UnsafeUtility.Malloc`, which gives you freedom to support Temp and TempJob allocation schemes... just have to remember to `UnsafeUtility.Free` after is the only issue.

ShadauxCat commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkVariableTests.cs:
```diff
@@ -4,8 +4,63 @@
+
 namespace Unity.Netcode.RuntimeTests
 {
+    public unsafe struct FixedStringStruct : INetworkSerializable
```
My overall question when it comes to fixed strings is... what's the use case? Any time you need to actually write a string to a log or pass it to a function or draw it to the screen... wouldn't you end up needing to convert it to string and thus incurring an allocation? So does it actually stop allocations or does it just delay them?

One thing I was going to experiment with was doing something like this:

```c#
struct FixedString
{
  public readonly string String;

  public FixedString(int size)
  {
    String = "".PadRight(size);
  }

  public string AsString()
  {
    return String;
  }

  public unsafe char this[int index]
  {
    get => m_Value[index];
    set
    {
      fixed(char* charPtr = String)
      {
        charPtr[index] = value;
      }
    }
  }
}
```

That would give you a mutable string - one-time allocation up front, but from there you're able to overwrite the memory of the string directly instead of creating new ones.

mattwalsh-unity replied:
This sketch came up because someone was asking how we do strings with NetworkVariables now that we restrict them to unmanaged types.  The answer is, especially as these vars get snapshotted and as we move toward "performant netcode / MTU-friendly oriented", we do not want arbitrary length types in netvars.  Yet I was interested in seeing what would happen if we used a fixed string scheme.  In that sense the pattern with a struct with a byte/char array of declared size is more what I'm trying to test / demo

NoelStephensUnity replied:
Oooo... I like your approach Jaedyn.  It only unsafely encapsulates when you are setting the value directly.  You padright call is a nice touch... to make sure it ends it properly.

NoelStephensUnity replied:
@ShadauxCat 
The fixed buffer allows you to re-use the pre-allocated char array.
So, if you were to have like a player name that could vary in size then you could have something like this defined in a class as a NetworkVariable<T> which would allow you to have a "placeholder" string that wasn't "nullable".

andrews-unity commented:
So we have a FixedString class in Collections... its tested, it worked, we use it for UTP today. 

https://docs.unity3d.com/Packages/com.unity.collections@1.0/api/Unity.Collections.FixedString.html

NoelStephensUnity committed:
refactor

minor improvement on fixed string test.

NoelStephensUnity committed:
refactor

minor improvement on fixed string test.

mattwalsh-unity committed:
Merge branch 'develop' into test/fixed-strings

mattwalsh-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkVariableTests.cs:
```diff
@@ -4,8 +4,76 @@
+
+        public void NetworkSerialize(NetworkSerializer serializer)
+        {
+            if (serializer.IsReading)
```
I want to come back and revisit this, just not today.  I also would like to re-write with `FixedString32` @andrews-unity , and I got it close but again, would like to get this in and improve later.

mattwalsh-unity committed:
Merge branch 'develop' into test/fixed-strings

NoelStephensUnity committed:
test

Testing FixedString32 in an INetworkSerializable struct.

NoelStephensUnity committed:
test: fixed strings and network variables

Testing FixedString32 in an INetworkSerializable struct.

NoelStephensUnity committed:
Merge branch 'test/fixed-strings' of https://github.com/Unity-Technologies/com.unity.netcode.gameobjects into test/fixed-strings

NoelStephensUnity committed:
fix

Fixing merge issue.

NoelStephensUnity committed:
fix

Fixing merge issue.

NoelStephensUnity committed:
test

touching the asmdef to see if this changes the issue with not being able to find the Unity.Collections assembly.

NoelStephensUnity committed:
test

touching the asmdef to see if this changes the issue with not being able to find the Unity.Collections assembly.

NoelStephensUnity commented:
> So we have a FixedString class in Collections... its tested, it works, we use it for UTP today.
> 
> https://docs.unity3d.com/Packages/com.unity.collections@1.0/api/Unity.Collections.FixedString.html

@andrews-unity  
So what is the trick to get Unity.Collections working in our runtime test assembly definition?  :)
We want to nix the other unsafe fixed string struct and run with FixedString32...and what not.

andrews-unity commented:
> > So we have a FixedString class in Collections... its tested, it works, we use it for UTP today.
> > https://docs.unity3d.com/Packages/com.unity.collections@1.0/api/Unity.Collections.FixedString.html
> 
> @andrews-unity
> So what is the trick to get Unity.Collections working in our runtime test assembly definition? :)
> We want to nix the other unsafe fixed string struct and run with FixedString32...and what not.

You should just be able to reference collections in the test assembly. I pulled down this branch and it seemed fine, and I was able to run the tests.

andrews-unity commented on com.unity.netcode.gameobjects/Tests/Runtime/NetworkVariableTests.cs:
```diff
@@ -3,9 +3,101 @@
+        public FixedString32 String;
+        public void NetworkSerialize(NetworkSerializer serializer)
+        {
+            if (serializer.IsReading)
```
Likely could make this even better when @ShadauxCat  stuff lands because then we can just write the length of the FixedString and then also write the raw byte* and then do the same for the read so there is no need to create the managed string.

NoelStephensUnity replied:
> > > So we have a FixedString class in Collections... its tested, it works, we use it for UTP today.
> > > https://docs.unity3d.com/Packages/com.unity.collections@1.0/api/Unity.Collections.FixedString.html
> > 
> > 
> > @andrews-unity
> > So what is the trick to get Unity.Collections working in our runtime test assembly definition? :)
> > We want to nix the other unsafe fixed string struct and run with FixedString32...and what not.
> 
> You should just be able to reference collections in the test assembly. I pulled down this branch and it seemed fine, and I was able to run the tests.

The reason this PR is failing is because that assembly can't find the Unity.Collections namespace.  It is failing on a compilation error.

andrews-unity replied:
Tests are green now :D the issue was with the fact that collections is not a direct dependency of the NGO package and so when the upm-ci does it isolation testing it won't find them. I added collections as a dependency

NoelStephensUnity replied:
OMG... Ok... so this package 
`"com.unity.collections": "0.12.0-preview.13"`
Wasn't in the packages... haha.  Ok, thanks for helping for sure!  (lol)
Will remember to check that next time this happens.

andrews-unity committed:
fix: NGO needs to directly depend on collections

Collections happened to be pulled in by UTP package but if we are testing NGO by itself in package testing then it needs to depend directly on collections.

andrews-unity committed:
fix: NGO needs to directly depend on collections

Collections happened to be pulled in by UTP package but if we are testing NGO by itself in package testing then it needs to depend directly on collections.

NoelStephensUnity committed:
Merge branch 'develop' into test/fixed-strings

NoelStephensUnity committed:
refactor

remove the unsafe related code

NoelStephensUnity committed:
refactor

remove the unsafe related code

mattwalsh-unity committed:
Merge branch 'develop' into test/fixed-strings

NoelStephensUnity reviewed (approved):
LGTM!

mattwalsh-unity committed:
Merge branch 'develop' into test/fixed-strings

mattwalsh-unity committed:
Merge branch 'develop' into test/fixed-strings

mattwalsh-unity committed:
Merge branch 'test/fixed-strings' of github.com:Unity-Technologies/com.unity.multiplayer.mlapi into test/fixed-strings

mattwalsh-unity committed:
variable naming cleanup

mattwalsh-unity committed:
test cleanup, allow updatePlayerPrefab to be null

mattwalsh-unity committed:
whitespace cleanup
